client/taskenv/env.go

```diff
@@ -290,6 +307,53 @@ func (t *TaskEnv) ReplaceEnv(arg string) string {
 	return hargs.ReplaceEnv(arg, t.EnvMap, t.NodeAttrs)
 }
 
+// replaceEnvClient takes an arg and replaces all occurrences of client-specific
+// environment variables and Nomad variables.  If the variable is found in the
+// passed map it is replaced, otherwise the original string is returned.
+// The difference from ReplaceEnv client is potentially different values for
+// the following variables:
+// * NOMAD_ALLOC_DIR
+// * NOMAD_TASK_DIR
+// * NOMAD_SECRETS_DIR
+// and anything that was interpolated using them.
+//
+// See https://github.com/hashicorp/nomad/pull/9671
+func (t *TaskEnv) replaceEnvClient(arg string) string {
+	return hargs.ReplaceEnv(arg, t.EnvMapClient, t.NodeAttrs)
+}
+
+// checkEscape returns true if the absolute path testPath escapes both the
+// task directory and shared allocation directory specified in the
+// directory path fields of this TaskEnv
+func (t *TaskEnv) checkEscape(testPath string) bool {
+	for _, p := range []string{t.clientTaskDir, t.clientSharedAllocDir} {
+		if p != "" && !helper.PathEscapesSandbox(p, testPath) {
+			return false
+		}
+	}
+	return true
+}
+
+// ClientPath interpolates the argument as a path, using the
+// environment variables with client-relative directories. The
+// result is an absolute path on the client filesystem.
+//
+// If the interpolated result is a relative path, it is made absolute
+// wrt to the task working directory.
+// If joinEscape, an interpolated path that escapes will be joined with the
+// task dir.
```

Comment:
Can you add an example usage here? Maybe it's the holidays, but I don't want to admit how long I stared at this and the relevant 2 lines of code to figure out what was going on. :sweat_smile:

Reply:
yeah, will do. this is the hiccup that i ran up against, where previous behavior allowed for absolute destination paths for `artifact` and `template`: 
https://github.com/hashicorp/nomad/pull/9671/files#diff-90e77cdb4f5439161bf5a053bf2d0b74d3ac3ee65cfce482bbd8e02646a14a64L140-L144

i'll add a documenting test.

Reply:
documenting test here: 
https://github.com/hashicorp/nomad/pull/9671/files#diff-6c8db153b63f572b0e0d0f447c3071b491c157fee3fe7c2c1040aaa70b40bf76R895-R903